src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt

```diff
@@ -25,29 +27,52 @@ class EditDream(
     dreamRequest: DreamRequest,
     userId: Long,
     currentTime: Long = System.currentTimeMillis()
-  ): Either<DreamEditionErrors, Dream> = dao.findById(dreamId)
-    .toOption()
-    .map { it.toDomain() }
-    .toEither { DreamEditionErrors.DreamNotFound(dreamId) }
-    .flatMap {
+  ): Either<DreamEditionErrors, Dream> = (
+    retrieveDreamFromPersistence(dao, dreamId) andThen
+      checkUserIsCreatorOfDream(userId) andThen
+      updateDream(dreamRequest, currentTime) andThen
+      validate(currentTime) andThen
+      storeDream(userId)
+    )()
+
```

Comment:
What about now? @tonilopezmr

Reply:
I'm interested on your opinions too @alvarocasadoc @Diphendara @kevinzs. Which way do you prefer? Do you understand this better now? We can work on making this even simpler but this is "functional" enough.

Reply:
Bump

Reply:
I didn't read any review for this code and I'm waiting for feedback 😢

Reply:
I talk about the question, but i don't have any to say about the code, looks fine for me.

Reply:
Do you prefer it as it is now or do you like it like it was before? 

Furthermore, we can even change it into two more ways: 

Do you prefer to remove the `andThen` operators and work with extension functions instead of higher order functions (functions returning lambdas)?

```kotlin
fun execute(
    dreamId: Long,
    dreamRequest: DreamRequest,
    userId: Long,
    currentTime: Long = System.currentTimeMillis()
  ): Either<DreamEditionErrors, Dream> = retrieveDreamFromPersistence(dreamId)
      .checkUserIsCreatorOfDream(userId)
      .updateDream(dreamRequest, currentTime)
      .validate(currentTime)
      .storeDream(userId)
```
This style is the least functional (but maybe it is the more pragmatic)

Or do you prefer using more advanced concepts of functional programming and don't write code like a stream and do sequential operations?

```kotlin
fun execute(
    dreamId: Long,
    dreamRequest: DreamRequest,
    userId: Long,
    currentTime: Long = System.currentTimeMillis()
  ): Either<DreamEditionErrors, Dream> = ForEither extensions {
  bindings {
   val dream = retrieveDreamFromPersistence(dreamId).bind()
   
   if (isUserCreatorOfDream(userId, dream)) {
     val updatedDream = updateDream(dream, dreamRequest, currentTime).bind()
     validateDream(updatedDream, currentTime).bind()
   } else {
     UserIsNotCreatorOfDream.left()
   }
  }.fix()
}
```
This is the `monad comprehension` style and it's very _haskelly_ but it introduces some noise and it's not as flexible as pure function composition.

Reply:
I like the 1º option of this comment, I can understand it and read more quickly

Reply:
More opinions on this please @alvarocasadoc @tonilopezmr @kevinzs @alexares @jodoldar @dredonieto

Reply:
Choosing between this two options I prefer the first. I didn't like the third, I prefer the original form instead of this :laughing: .